### ISSUE TITLE
enable export to Geant4 text geometry description 

### ISSUE THREAD

**Issue by jintonic** [jintonic/shine#12]
General information about Geant4 text geometry description: <https://github.com/jintonic/gears/tree/master/tutorials/detector>

**Comment by bestdeveloper23** [comment]
Yes there is a way to display text to describe an object.
I will choose a reasonable method.

**Comment by jintonic** [comment]
No. "Geant4 text geometry description" is a format to store geometry information in a file.

**Comment by bestdeveloper23** [comment]
Oh, sorry.
I checked the sample code.
I understand what you mean.
So should I import .gdml file?

**Comment by bestdeveloper23** [comment]
#12 and #14 errors are opposite processes.
I will check the possiblities.

**Comment by jintonic** [comment]
The importing of GDML should be easy, because there is existing code.
The exporting of GDML needs some work, but we can learn from the importing code.
The exporting of .tg has to start from scratch, nobody did anything before. We can start with exporting a cube first and then expand to other shapes.

**Comment by bestdeveloper23** [comment]
Yes. you are right.
Thee.js has the functionality of importing of GDML file.
But the only way to upload .tg files is through Geant4.
So we need to create module to import this .tg file in our project from scratch.

**Comment by jintonic** [comment]
No. We don't need to IMPORT .tg file into "shine". We just need to EXPORT geometries in "shine" to a .tg file.

**Comment by bestdeveloper23** [comment]
After exporting, there is no need to import it again?

**Comment by jintonic** [comment]
We can do that in the future if it is really needed. At this moment, we don't need it.

**Comment by bestdeveloper23** [comment]
Okay. Let's do it in the future.
When can we start working in earnest?

**Comment by jintonic** [comment]
We need to agree with a rough estimation of the hours you need to finish these tasks.

**Comment by bestdeveloper23** [comment]
Have you agreed to start working at university?
And from now on, it will take more time than before because we have to create a completely new feature that has no precedent.
I need to study the previous examples and think a lot based on them.
I think it will take more time to think and test than to write code.
What do you think?

**Comment by jintonic** [comment]
I agree. Here is my estimation:

- importing GDML, 1 hour of coding, 1 hour of testing, 1 hour of evaluation. $90 in total. This is because we already have existing code. We just need to integrate it with our code.
- adding new examples, such as a skull, kidney, etc. 1 hour for each example. This is because we can find online many existing models in stl or other format. We just need to simplify them a bit.
- export to .tg format. 
  - 1 hour, export an empty scene to a 10x10x10 meter^3 experimental hall filled with air in a tg file. This is basically copy and paste an existing exporting function.
  - 1 hour, export a box to .tg file
  - 1 hour, export a cylinder to .tg file
  - 1 hour, export a sphere to .tg file
- export to GDML format.
  - 1 hour, find code to export in xml file
  - 1 hour, export an empty scene to a 10x10x10 meter^3 experimental hall filled with air in a gdml file.
  - 1 hour, export a box to gdml file

**Comment by bestdeveloper23** [comment]
1. I agree with that.
2. Should we have examples as assets?
3. To implement .tg file export, you need to learn all the features of .tg files.
I'm not sure how long it will take.
Where can I find existing code?
4. 
Where can I find existing code?

**Comment by jintonic** [comment]
Yes, examples can be included as assets. I want to make sure that the size is less than 5 MB for each example.

We can add features to .tg one by one. The first one (adding a box) is very simple:
```
:volume name_of_volume BOX half_x half_y half_z G4_material
```

You already implemented all G4_material. You also added all Geant4 basic shapes, `G4Box` translates to `BOX`, so on and so forth.

There is no existing code for tg exporting. But three.js editor also has a lot of built-in exporters. We can borrow the code to implement .tg exporter.

**Comment by bestdeveloper23** [comment]
Yes. I can use exporters, exactly ObjExporter.
But I think that for exporting of .tg file, I need to convert scene to tg texts.
Anyway, as a draft I agree with your suggestions.
I think it is fundamental to start work quickly and make progress.

**Comment by bestdeveloper23** [comment]
I am in the process of registering for level 3.
However, there are quite a few fields required.

**Comment by jintonic** [comment]
- [ ] You need to convert a "scene" in three.js to "world" in geant4 text geometry, which can be just a large box with a dimension of 10 m x 10 m x 10 m
- [ ] You can then convert a box in the scene to a box in geant4 text geometry with the same dimension and location in the world.

**Comment by bestdeveloper23** [comment]
Should I add the unit?
https://monosnap.com/file/LUIR4ZUwWrAZZVdBw6Gm09cploYqGK

**Comment by bestdeveloper23** [comment]
https://monosnap.com/file/kG4qnMjAGflhCAjAZfKIPk1iv4EfIo

I did

**Comment by jintonic** [comment]
Thank you so much! That looks great. One small thing though: the rotation matrix r000 needs to be read from three.js as well. The matrix can be defined in different ways. The definition can be found in page 15 of https://geant4.web.cern.ch/collaboration/working_groups/persistency/docs/textgeom.pdf

**Comment by bestdeveloper23** [comment]
https://monosnap.com/file/NmAPxrhy9hKfBxq33I54d3uq0TvBWN
I fixed.
Exporting of GDML is next?

**Comment by jintonic** [comment]
Great! Yes, please move forward to GDML exporting. How many hours did you work on text geometry exporting?

**Comment by bestdeveloper23** [comment]
2 hours yesterday, 1/2 hour today

I need a lot of work at once.
I think I should create these templates and then create them for all your models.
With your permission, I will try to implement it for other models after completing the GDML export template.
So I want to be able to work full time.
Please let me know your opinion.

**Comment by jintonic** [comment]
That's what I am thinking as well. I agree with you. Please go ahead with your plan. I should be able to support you to work full time for a while. But remember, this job alone cannot give you long term financial stability.

**Comment by bestdeveloper23** [comment]
I see.
And should be able to export the entire scene(not each model)?
Now we can export one selected model

**Comment by bestdeveloper23** [comment]
All right.

**Comment by jintonic** [comment]
The idea is for the entire scene (or a few selected geometries).

**Comment by bestdeveloper23** [comment]
I'll do some more research. First, we will implement an export function for all models.

**Comment by jintonic** [comment]
Yes, we need one function for each geometry model. But as you planned, we can work on individual models later. Right now, we can work on GDML exporting template.

**Comment by bestdeveloper23** [comment]
Yes. I got it.

**Comment by bestdeveloper23** [comment]
https://monosnap.com/file/rwGTt7SplRum7CTWVyP30M93ttfili

I exported gdml box.

**Comment by bestdeveloper23** [comment]
Where can I find all tg file example of models?
Because I need the format of configuration.

**Comment by jintonic** [comment]
How many hours did you use to enable exporting a box in gdml format?

**Comment by bestdeveloper23** [comment]
3 hours.
In fact, it initially takes 2 to 3 hours to understand the structure of the entire file.

**Comment by bestdeveloper23** [comment]
https://monosnap.com/file/focpp15ZJpgxGBrpWIXN6h9txLryl1

**Comment by jintonic** [comment]
For our record, you have worked for 16 hours now (4 h for GDML import, 4 h for examples, 2.5 h for tg export and 5.5 hours for GDML export).

**Comment by bestdeveloper23** [comment]
Okay.
https://monosnap.com/file/focpp15ZJpgxGBrpWIXN6h9txLryl1
Please provide the documentation, so I can refer.

**Comment by bestdeveloper23** [comment]
I added SPhere and Tubs.
But I can't do other models because I don't know the Name of model in tg and gdml.

**Comment by bestdeveloper23** [comment]
Do you understand what I'm asking?

**Comment by jintonic** [comment]
text geometry shape list starts from page 4 of https://geant4.web.cern.ch/collaboration/working_groups/persistency/docs/textgeom.pdf

**Comment by bestdeveloper23** [comment]
https://monosnap.com/file/dZihw7liwvVHXJFiSeDmmMUyU18nqB
This TUB doesn't have pSPhi and pDPhi.
Only need 3 parameters. 
Is it a full circle?

**Comment by jintonic** [comment]
`TUB` should go for a full circle. `TUBS` has pSPhi (starting phi angle) and pDPhi (delta angle of the segment)

**Comment by bestdeveloper23** [comment]
https://monosnap.com/file/QNYQj987jIB8VPxBjvycQbI4aQpb8e
I am not sure these parameter because it different from 
https://geant4-userdoc.web.cern.ch/UsersGuides/ForApplicationDeveloper/html/Detector/Geometry/geomSolids.html#constructed-solid-geometry-csg-solids

**Comment by bestdeveloper23** [comment]
https://monosnap.com/file/sxFjw7Ep6LlAZMndGLcYtiDPTtHL3U
I think the Delta angle of the segment is missed.

**Comment by jintonic** [comment]
Good observation! I think the tg manual is wrong! Just follow the order of parameters specified in Geant4 manual and ignore the tg manual if it's different from the Geant4 manual.

**Comment by bestdeveloper23** [comment]
I see.
Thank you.

**Comment by bestdeveloper23** [comment]
https://monosnap.com/file/luVWpWMy19Gsq93wiEhitusZzBOzrZ
This parabolid is missed.

**Comment by bestdeveloper23** [comment]
I updated code.
8h-
       macro file
       some tg models
8h-
       rest tg models
       some gdml exports
       
I would appreciate it if you would pay $500 now.

**Comment by jintonic** [comment]
All right. You worked for 32 hours in total. 17 hours are covered by my previous payment. I'll pay you $1,020 for 34 hours of work. You already worked for 32 - 17 = 15 hours. So 34 - 15 = 19 hours are pre-paid. Please let me know if you don't agree.

**Comment by bestdeveloper23** [comment]
I agree.
Thank you.

**Comment by bestdeveloper23** [comment]
Actually, yesterday was tax day, so I really needed money.
I would really appreciate it.
And I worked gdml models for 8 hours.
Next, I will implement full scene export for tg and gdml.
Okay?

**Comment by jintonic** [comment]
Sure. I just want to confirm that you received my prepayment of $1,020 through PayPal yesterday.

**Comment by bestdeveloper23** [comment]
Yes. I received.

**Comment by jintonic** [comment]
Great. So you already worked 8 hours for GDML import. Is that right?

**Comment by jintonic** [comment]
Thank you for your hard work! I just looked at your code quickly. It seems that you always assign "Air" as the material to your GDML shapes. Ideally, each three.js shape should have a material associated. That material should be exported to GDML together with the shape. The list of material is https://geant4-userdoc.web.cern.ch/UsersGuides/ForApplicationDeveloper/html/Appendix/materialNames.html and I remember you already implemented.

**Comment by bestdeveloper23** [comment]
https://monosnap.com/file/eLZ958IbzxiFI5BaXZTKwHz2RL5suv

You can check again.
I already implemented that. :)

**Comment by jintonic** [comment]
Cool! Thank you!

**Comment by bestdeveloper23** [comment]
I sometimes study nuclear physics to make our work more efficiently.
It's true

**Comment by jintonic** [comment]
Wow! I appreciate your dedication to this work!

**Comment by bestdeveloper23** [comment]
https://monosnap.com/file/1iRH9e3zoufdvcRW2L51L8k0TlijWP
I completed scene.tg export.

**Comment by jintonic** [comment]
I enjoyed watching your screencast :)
How many hours did you spend on this?
- [ ] the name of the volume should be taken from the "Name" field in the "OBJECT" tab in the right panel.
- [ ] If we put two solids in the scene with the same shape, there should be a way to make sure that their solid names in the tg file are different.
- [ ] As shown in the screenshot below, the `:volu` command should take the name of the solid from the `:solid` command. This is how a volume is connected with a solid.
![image](https://github.com/jintonic/shine/assets/4230658/df737472-5c16-44b8-bcde-7edd8f763120)

**Comment by jintonic** [comment]
OK, you worked on GDML for 8 h, on TG for 12 h. I will pay you another $1,020 for 34 h of work. You already worked 1 more hour without pay. After I pay you another $1,020, you can work for 33 h. Is that right?

**Comment by bestdeveloper23** [comment]
1. I used the "Name" field in the "OBJECT" tab.
https://monosnap.com/file/Q1XoSpgIvGMOmLzcCszl7a9HbpY9W5
2. To give different names to the same shape, suffixes were added.
  https://monosnap.com/file/ovZcesY0V3GxRI5El0czdqwYSlJ9Or
  
3. I need to fix that.

**Comment by bestdeveloper23** [comment]
Or do you need each index number for each shape?
https://monosnap.com/file/FVx89lCJQU0UQpalaQHIAFrcvs2Br9

Like this?

**Comment by bestdeveloper23** [comment]
Thank you.
I received.

**Comment by bestdeveloper23** [comment]
https://monosnap.com/file/wqvSR5TmOp4fLMxqHMGOAMq4HmO5mQ
gdml completed.
I worked for 9 hours.
Please check and give me feedback.
And please give me other task to do while you write feedback.

**Comment by bestdeveloper23** [comment]
I implemented export boolean object with tg file.
https://monosnap.com/file/cPFsT2WfrCsxcR6ib1ViuKn8mA43Nb
I worked for 18 hours for this.

**Comment by bestdeveloper23** [comment]
yes. Last weekend I worked 9 hours, and this week I have worked 18 hours so far.
Please check the details carefully.
I think I implemented it correctly.
All conditions were taken into consideration.
Thank you

**Comment by bestdeveloper23** [comment]
I implemented boolean object export with .gdml.
https://monosnap.com/file/4r1qr2TqJjvHvURGYGTUvstJKBG8XU

**Comment by jintonic** [comment]
That's exciting. How many hours did you work on this?

**Comment by bestdeveloper23** [comment]
I worked 8 hours for it.
And the exporting of scene including boolean object is almost complete.

**Comment by bestdeveloper23** [comment]
And this completed the export of the scene, including the Boolean objects.
I worked 6 hours for this.

**Comment by jintonic** [comment]
I'll pay you another $1020 for 34 hours, but you already worked for 8 more hours, so there will be 34 - 8 = 26 hours left. Please slow down. I need some time to check the current status.

**Comment by jintonic** [comment]
Feel free to check in the code. You don't have to wait for me.

**Comment by bestdeveloper23** [comment]
Yes. I see. 
Please check it quickly and let me know if there are any changes to be made.
If you go too far, it may take more time to correct the previous ones.

**Comment by jintonic** [comment]
Agreed. I'll find sometime to work on this.
